node/src/accountant/scanners.rs

```
@@ -0,0 +1,1871 @@
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NullScanner {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+pub struct ScannerMock<BeginMessage, EndMessage> {
+    begin_scan_params: Arc<Mutex<Vec<()>>>,
+    begin_scan_results: RefCell<Vec<Result<BeginMessage, BeginScanError>>>,
+    end_scan_params: Arc<Mutex<Vec<EndMessage>>>,
+    end_scan_results: RefCell<Vec<Option<NodeToUiMessage>>>,
+    stop_system_after_last_message: RefCell<bool>,
+}
+
+impl<BeginMessage, EndMessage> Scanner<BeginMessage, EndMessage>
+    for ScannerMock<BeginMessage, EndMessage>
+where
+    BeginMessage: Message,
+    EndMessage: Message,
+{
+    fn begin_scan(
+        &mut self,
+        _timestamp: SystemTime,
+        _response_skeleton_opt: Option<ResponseSkeleton>,
+        _logger: &Logger,
+    ) -> Result<BeginMessage, BeginScanError> {
+        self.begin_scan_params.lock().unwrap().push(());
+        if self.is_allowed_to_stop_the_system() && self.is_last_message() {
+            System::current().stop();
+        }
+        self.begin_scan_results.borrow_mut().remove(0)
+    }
+
+    fn finish_scan(&mut self, message: EndMessage, _logger: &Logger) -> Option<NodeToUiMessage> {
+        self.end_scan_params.lock().unwrap().push(message);
+        if self.is_allowed_to_stop_the_system() && self.is_last_message() {
+            System::current().stop();
+        }
+        self.end_scan_results.borrow_mut().remove(0)
+    }
+
+    fn scan_started_at(&self) -> Option<SystemTime> {
+        unimplemented!()
+    }
+
+    fn mark_as_started(&mut self, _timestamp: SystemTime) {
+        unimplemented!()
+    }
+
+    fn mark_as_ended(&mut self, _logger: &Logger) {
+        unimplemented!()
+    }
+}
+
+impl<BeginMessage, EndMessage> Default for ScannerMock<BeginMessage, EndMessage> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<BeginMessage, EndMessage> ScannerMock<BeginMessage, EndMessage> {
+    pub fn new() -> Self {
```

Again, this is a conflict of different "schools", but my preferences are to use the #[derive(Default)] attribute which will automatically hide a lot of unimportant lines. The default values for mocks are pretty predictable (because all these values implement Default too).